extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala

```diff
@@ -182,4 +194,82 @@ private[kinesis] class KinesisReceiver(
         new DefaultAWSCredentialsProviderChain()
     }
   }
+
+  /** Remember the range of sequence numbers that was added to the currently active block */
+  private def rememberAddedRange(range: SequenceNumberRange): Unit = {
+    seqNumRangesInCurrentBlock += range
+  }
+
+  /**
+   * Finalize the ranges added to the block that was active
+   * and prepare the ranges buffer for next block.
+   */
+  private def finalizeRangesForCurrentBlock(blockId: StreamBlockId): Unit = {
+    blockIdToSeqNumRanges(blockId) = seqNumRangesInCurrentBlock.toArray
+    seqNumRangesInCurrentBlock.clear()
+    logDebug(s"Generated block $blockId has ${ blockIdToSeqNumRanges(blockId).mkString(", ")}")
+  }
+
+  private def storeBlockWithRanges(
```

Comment:
Good catch. How about making the `blockIdToSeqNumRanges` implement `SynchronizedMap`?

The reason I dont want to add synchronized before called onPushBlock is because its a bad idea to synchronized long-running block of code (Receiver.store() can take a long time to return from storing the block). The storing of the block really does not need be done under a lock, only the accessing and updating of `blockIdToSeqNumRanges` needs to be done in a synchronized manner.